modules/openapi-generator/src/main/java/org/openapitools/codegen/utils/ModelUtils.java

```diff
@@ -1101,24 +1101,55 @@ public static Schema unaliasSchema(OpenAPI openAPI,
                         return schema; // generate a model extending map
                     } else {
                         // treat it as a typical map
-                        return unaliasSchema(openAPI, allSchemas.get(ModelUtils.getSimpleRef(schema.get$ref())),
+                        return unaliasSchema(openAPI, allSchemas.get(getSimpleRef(schema.get$ref())),
                                 importMappings);
                     }
                 }
             } else if (isObjectSchema(ref)) { // model
                 if (ref.getProperties() != null && !ref.getProperties().isEmpty()) { // has at least one property
                     return schema;
-                } else { // free form object (type: object)
-                    return unaliasSchema(openAPI, allSchemas.get(ModelUtils.getSimpleRef(schema.get$ref())),
+                } else {
+                    // free form object (type: object)
+                    if (hasValidation(ref)) {
+                        return schema;
+                    } else if (hasDescendants(simpleRef, openAPI)) {
+                        return schema;
```

Comment:
@jimschubert this will turn on interface schemas on everywhere
Is that what we want?

Reply:
@spacether I don't think this will "turn on interface models"… it's just returning a model that has no properties, which is a perfectly valid use case. We should have some way to indicate that the schema is free form, but we currently only store that boolean on response, parameter, and property codegen types. 

One question I would have is whether we have sufficient test coverage on this code path, because only a single typescript file has changed. For example, does `ModelUtils.isFreeFormObject` also have to be updated with logic around validation and descendants?

Reply:
The reason that I think that this will turn on interface models everywhere is that this is [how python-experimental turned it on](https://github.com/OpenAPITools/openapi-generator/commit/35d42728d9df60da65300fe7cd900e4056719969) and this change applies to all generators that use ModelUtils.unaliasSchema.
If you look [here](https://github.com/OpenAPITools/openapi-generator/blob/master/modules/openapi-generator/src/main/java/org/openapitools/codegen/DefaultGenerator.java#L463)
you see that when generating models, for a free form model, we attempt to unalias a ref to it and:
- if we follow the reference and return the schema that we ref point to, then we skip generating the model
- if we return the ref schema, we DO generate the model
This code makes it so we return the ref schema, so the interface model will be generated.
One way to confirm this would be to update the v2 and v3 petstore sample to include an interface model and see if it is generated.

In my opinion these interface models are free form models. That's why I included the comments:
```
                    // there are 3 free form use cases
                    // 1. free form with no validation that is not allOf included in any composed schemas
                    // 2. free form with validation
                    // 3. free form that is allOf included in any composed schemas
                    //      this use case arises when using interface schemas
                    // generators may choose to make models for use case 2 + 3
```

Reply:
What are your suggested updates to ModelUtils.isFreeFormObject @jimschubert?

Reply:
I'm not sure to understand but the case is `AbstractPet` must be both (in Java) a Class and a Map<String,String> (FreeForm)?
For the test coverage, I can add `composed-schemas` on some other generators if it's required.

Reply:
@spacether I think you're mixing the concepts of "free form object" and interfaces. These aren't exactly the same things. An interface is something that's used as a base for other instances, while a "free form object" is one which is defined as having an unbounded set of allowed properties. The current implementation before this pull request doesn't allow marker interfaces or free form objects in composed models, although these are supported by the specification. If python has issues with supporting the spec, I think the changes will be needed in python so as not to block spec support in everything else.

> What are your suggested updates to ModelUtils.isFreeFormObject

If we're going to return the schema for the two new cases in this method (for block labeled as "free form objects"), then I think we should have the same two conditionals in the `ModelUtils.isFreeFormObject` method. To put it differently, this method is now:

```
if (ref.getProperties() != null && !ref.getProperties().isEmpty()) { // has at least one property
    return schema;
} else {
    // free form object (type: object)
    if (hasValidation(ref)) {
        return schema;
    } else if (hasDescendants(simpleRef, openAPI)) {
        return schema;
    }
    return unaliasSchema(openAPI, allSchemas.get(getSimpleRef(schema.get$ref())),
            importMappings);
}
```

if the else block is still correct that we have free form objects in three different states now, then I think the helper method needs to be updated.

If, instead, the logic is meant to be this:

```
// return schema when it has at least one property, validations, or descendants
if (ref.getProperties() != null && !ref.getProperties().isEmpty()) {
    return schema;
}
if (hasValidation(ref)) {
    return schema;
} 
if (hasDescendants(simpleRef, openAPI)) {
    return schema;
}

// free form object (type: object)
return unaliasSchema(openAPI, allSchemas.get(getSimpleRef(schema.get$ref())),
        importMappings);
```

… then I think comments and structure might have to be updated in this block for clarity and `ModelUtils.isFreeFormObject` still makes sense as-is. Again, I question the amount of coverage we have on these scenarios because this PR hasn't changed any tests or samples which would indicate it's "turned on interface models".

Reply:
> If python has issues with supporting the spec, I think the changes will be needed in python so as not to block spec support in everything else.
Python-experimental supports all of these use cases and is not blocked.

I have lots of questions here. Are you free to discuss it in Slack?

1. How you would classify these use cases, InterfaceModel, ModelWithValidations
InterfaceModel:
  type: object
ModelWithValidations:
  type: object
  maxProperties: 3
ComposedSchema:
  allOf
"$ref": "#/components/schemas/ErrorModel"
2. Should ModelUtils.isMapSchema be true for these?
3. Can a free form object be used as an interface?
4. Doesn't a schema type:object that is allOf included in other schemas constrain the types of properties to be AnyType because additionalProperties={} by default? That is more constraining than an empty class definition (interface). These languages have many types not covered by the openapi spec.
5. Must those other instances that inherit from the interface be type object or can they be primitives like string with validations? If so then  aren't we talking about AnyType (type=null)?

Reply:
Does the discussion has occurred? :)